Yoast/wordpress-seo
# P1-1039 google preview container

## Context


* Wrap the Google preview in a container so that it already has the data from the SEO integration available

## Summary


This PR can be summarized in the following changelog entry:

* Adds a `GooglePreviewContainer`.

## Relevant technical choices:

### SEO integration
* Introduced a provider to pass non-store context to our containers within the SEO integration package. Used right away for the replacement variables. They include the `getReplacement` functions, which should not live in a Redux store. Kept the naming open ended so that we have one provider to rule them all 😉 
* Added a cache for the replacement variables for the Google preview that looks at the actual values every render. This is needed to ensure that we do refresh when a value changes. While at the same time does not update the editor state within the Snippet Editor, which causes a reset of the selection state (puts the cursor at the start of the input). Keeping the friction with our Snippet Editor component at a minimum.
* Ignoring any data that we do not handle (yet) inside our integration, like the shopping data. One should simply pass those props to the container.
* Providing an `as` prop to determine what the component actually is. There is not default of the SnippetEditor component right now. This is due to the `@yoast/seo-integration`'s Babel build not being compatible with the unbuild code from the `@yoast/search-metadata-previews` package. Decided to leave this problem to the implementor, since there we already have it available anyway. To tackle this in the future we could go to WP scripts instead of Babel, when they upgrade webpack to v5, so we could use the module build that includes the SnippetEditor, maybe.

### Example app
* Added Yoast packages to the webpack config, so we build our unbuild packages. As well as enforce the usage of the monorepo version, instead of relying on the published versions (as those are out-of-date).
* Include the draft-js CSS for the Snippet Editor without adding it as a specific dependency... is this wrong? 😇 
* Removed the replacement variables for the Google preview form fields, this caused updates all the time and are not really replacement variables anyway.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Install / build steps
* Install dependencies in the monorepo
* Link dependencies inside `@yoast/seo-integration`:
  * `@yoast/replacement-variables`
  * `@yoast/seo-store`
* Build the packages:
  * `@yoast/replacement-variables`
  * `@yoast/seo-store`
  * `@yoast/seo-integration`
* Install dependencies in the `@yoast/seo-integration-app` (note the app part here)
* Link `@yoast/seo-integration` inside the app
* Start the app

#### Test
* Verify you have the Snippet editor in the sidebar
* Verify you can change the fields
* Verify you can use the replacement variables (via typing: `%%title%%` and via the insert button)
* Verify the preview shows the replaced values
* Change the date, verify it shows up in the preview
* Change the permalink, verify it changes the breadcrumb in the preview
* Change the mode, verify it switches just fine
* Did I miss any data that could be included already?
* Bonus points: add some props to the implementation in the app (like `shoppingData`) and verify it gets passed to the component

### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

* N/A

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes


## Other environments


## Documentation


## Quality assurance


Fixes https://yoast.atlassian.net/browse/P1-1039
